implement queue system with Redis integration and add queue routes

dori654 opened the pull request:
not all working yet, I need feedback to proceed.

dori654 committed:
implement queue system with Redis integration and add queue routes

not all working yet, I need feedback to proceed.

ShaySarussiElshten reviewed (commented):
its look great implmantion 👍 you could procced to do the other task in this epic.

dont forgrt switch to master , pull the change , and merge it to your branch.
(i merge pr to master)

ShaySarussiElshten reviewed (commented):
why all the time i got in container of backend ?

🔄 Processing job 785 with data: { task: 'load-task-783' }
app-1    | 🛠️ Job 785 is now active.
app-1    | 🔄 Processing job 784 with data: { task: 'load-task-782' }
app-1    | 🛠️ Job 784 is now active.
app-1    | 🔄 Processing job 783 with data: { task: 'load-task-781' }
app-1    | 🛠️ Job 783 is now active.
app-1    | 🔄 Processing job 782 with data: { task: 'load-task-780' }
app-1    | 🛠️ Job 782 is now active.
app-1    | ✅ Job 785 completed successfully.
app-1    | ✅ Job 784 completed successfully.
app-1    | ✅ Job 783 completed successfully.
app-1    | ✅ Job 782 completed successfully.
app-1    | ✅ Job 785 successfully completed.
app-1    | ✅ Job 784 successfully completed.
app-1    | ✅ Job 783 successfully completed.
app-1    | 🔄 Processing job 781 with data: { task: 'load-task-779' }
app-1    | 🛠️ Job 781 is now active.
app-1    | 🔄 Processing job 780 with data: { task: 'load-task-778' }
app-1    | 🛠️ Job 780 is now active.
app-1    | 🔄 Processing job 779 with data: { task: 'load-task-777' }
app-1    | 🛠️ Job 779 is now active.
app-1    | ✅ Job 782 successfully completed.
app-1    | 🔄 Processing job 778 with data: { task: 'load-task-776' }
app-1    | 🛠️ Job 778 is now active.
app-1    | ✅ Job 781 completed successfully.
app-1    | ✅ Job 780 completed successfully.
app-1    | ✅ Job 779 completed successfully.
app-1    | ✅ Job 778 completed successfully.
app-1    | ✅ Job 781 successfully completed.
app-1    | ✅ Job 780 successfully completed.
app-1    | ✅ Job 779 successfully completed.
app-1    | 🔄 Processing job 777 with data: { task: 'load-task-775' }
app-1    | 🛠️ Job 777 is now active.
app-1    | 🔄 Processing job 776 with data: { task: 'load-task-774' }
app-1    | 🛠️ Job 776 is now active.
app-1    | 🔄 Processing job 775 with data: { task: 'load-task-773' }
app-1    | 🛠️ Job 775 is now active.
app-1    | ✅ Job 778 successfully completed.
app-1    | 🔄 Processing job 774 with data: { task: 'load-task-772' }
app-1    | 🛠️ Job 774 is now active.



i need ecxcute this only in req not all the time in background !

ShaySarussiElshten reviewed (commented):
the test in not pass i got error like:

 🔄 Processing job 1003 with data: { task: 'load-task-0' }

      at QueueService.log (src/services/queue.service.ts:101:17)

  console.log
    🛠️ Job 1003 is now active.

      at Queue.log (src/services/queue.service.ts:136:15)

  console.log
    🔄 Processing job 1004 with data: { task: 'load-task-1' }

      at QueueService.log (src/services/queue.service.ts:101:17)

  console.log
    🛠️ Job 1004 is now active.

      at Queue.log (src/services/queue.service.ts:136:15)

  console.log
    🔄 Processing job 1745 with data: { task: 'load-task-742' }

      at QueueService.log (src/services/queue.service.ts:101:17)

  console.log
    🛠️ Job 1745 is now active.

      at Queue.log (src/services/queue.service.ts:136:15)

2025-02-01T18:30:15.440Z [App] info: Server stopped {"service":"App"}
2025-02-01T18:30:15.461Z [App] info: Server stopped {"service":"App"}

  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "🔄 Processing job 1005 with data: { task: 'load-task-2' }".

       99 |     this.queue.process(config.concurrency, async (job) => {
      100 |       try {
    > 101 |         console.log(`🔄 Processing job ${job.id} with data:`, job.data);
          |                 ^
      102 |
      103 |         // Simulated processing delay
      104 |         await new Promise(resolve => setTimeout(resolve, 2000));

      at console.log (node_modules/@jest/console/build/CustomConsole.js:141:10)
      at QueueService.log (src/services/queue.service.ts:101:17)
      at src/services/queue.service.ts:3273:41
      at Object.<anonymous>.__awaiter (src/services/queue.service.ts:3222:10)
      at Queue.__awaiter (src/services/queue.service.ts:99:58)
      at handlers.<computed> (node_modules/bull/lib/queue.js:733:42)
      at Queue.Object.<anonymous>.Queue.processJob (node_modules/bull/lib/queue.js:1210:22)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "🛠️ Job 1005 is now active.".

      134 |
      135 |     this.queue.on('active', (job) => {
    > 136 |       console.log(`🛠️ Job ${job.id} is now active.`);
          |               ^
      137 |     });
      138 |
      139 |     this.queue.on('removed', (job) => {

      at console.log (node_modules/@jest/console/build/CustomConsole.js:141:10)
      at Queue.log (src/services/queue.service.ts:136:15)
      at Object.<anonymous>.module.exports.emitSafe (node_modules/bull/lib/utils.js:50:20)
      at Queue.Object.<anonymous>.Queue.processJob (node_modules/bull/lib/queue.js:1217:11)

 PASS  test/unit/health.test.ts (6.494 s)
  Health Check API
    GET /health
      ✓ should handle valid request (37 ms)
      ✓ should handle request without api key (21 ms)

 FAIL  test/unit/queue.test.ts (6.56 s)
  Queue System Tests
    ✕ ✅ Should enqueue a job (43 ms)
    ✓ ✅ Should retrieve job status (18 ms)
    ✕ ✅ Should retry a failed job (1 ms)
    ✓ ✅ Should cancel a queued job (6 ms)
    ✕ ✅ Should return 404 when retrying a non-existent job (8 ms)
    ✓ ✅ Should clear the queue (5 ms)
    ✓ ✅ Should handle 1000 concurrent job enqueues (Load Test) (133 ms)
    ✕ ✅ Should process jobs without failure (6 ms)

  ● Queue System Tests › ✅ Should enqueue a job

    listen EADDRINUSE: address already in use :::3000

      62 |       }
      63 |
    > 64 |       this.server = this.app.listen(config.port, () => {
         |                              ^
      65 |         logger.info(`Server started on port ${config.port}`);
      66 |         logger.info(`Health check endpoint: ${config.baseUrl}/health`);
      67 |         logger.info(`Queue system is running : ${config.baseUrl}/queue`);

      at Function.listen (node_modules/express/lib/application.js:635:24)
      at App.listen (src/app.ts:64:30)
      at src/app.ts:2790:40
      at Object.<anonymous>.__awaiter (src/app.ts:2739:10)
      at App.start (src/app.ts:2955:12)
      at test/unit/queue.test.ts:16:21
      at test/unit/queue.test.ts:8:71
      at Object.<anonymous>.__awaiter (test/unit/queue.test.ts:4:12)
      at Object.<anonymous> (test/unit/queue.test.ts:12:22)

  ● Queue System Tests › ✅ Should retry a failed job

    Couldn't retry job: The job has been already retried or has not failed

      at node_modules/bull/lib/job.js:387:15

  ● Queue System Tests › ✅ Should return 404 when retrying a non-existent job

    expect(received).toBe(expected) // Object.is equality

    Expected: 404
    Received: 200

      60 |       .post(`/api/queue/retry/invalid-job-id`);
      61 |
    > 62 |     expect(response.status).toBe(404);
         |                             ^
      63 |     expect(response.body.message).toBe('Job not found');
      64 |   });
      65 |

      at test/unit/queue.test.ts:62:29
      at fulfilled (test/unit/queue.test.ts:5:58)

  ● Queue System Tests › ✅ Should process jobs without failure

    expect(received).toBeLessThan(expected)

    Expected: < 10
    Received:   12

      82 |   it('✅ Should process jobs without failure', async () => {
      83 |     const activeJobs = await queueService['queue'].getActive();
    > 84 |     expect(activeJobs.length).toBeLessThan(10); // Controlled concurrency
         |                               ^
      85 |   });
      86 | });
      87 |

      at test/unit/queue.test.ts:84:31
      at fulfilled (test/unit/queue.test.ts:5:58)

dori654 committed:
refactor queue controller and tests: changed bull to bullMQ , improve error handling, add job status checks, and enhance test coverage

dori654 committed:
Merge branch 'master' of https://github.com/InnovatorsHub1/backend into BE-33

ShaySarussiElshten commented:
hey please resolve the conflict 👍

dori654 committed:
resolve conflicts

dori654 committed:
Merge branch 'master' of https://github.com/InnovatorsHub1/backend into BE-33

ShaySarussiElshten commented:
again test not runing 


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "⏳ Job 3850 is waiting to be processed.".

      156 |
      157 |     this.queueEvents.on('waiting', ({ jobId }) => {
    > 158 |       console.log(`⏳ Job ${jobId} is waiting to be processed.`);
          |               ^
      159 |     });
      160 |
      161 |     this.queueEvents.on('progress', ({ jobId, data }) => {

      at console.log (node_modules/@jest/console/build/CustomConsole.js:141:10)
      at QueueEvents.log (src/services/queue.service.ts:158:15)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-base.ts:121:20)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-events.ts:226:18)
      at QueueEvents.consumeEvents (node_modules/bullmq/src/classes/queue-events.ts:318:18)
      at QueueEvents.run (node_modules/bullmq/src/classes/queue-events.ts:272:9)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "⏳ Job 3851 is waiting to be processed.".

      156 |
      157 |     this.queueEvents.on('waiting', ({ jobId }) => {
    > 158 |       console.log(`⏳ Job ${jobId} is waiting to be processed.`);
          |               ^
      159 |     });
      160 |
      161 |     this.queueEvents.on('progress', ({ jobId, data }) => {

      at console.log (node_modules/@jest/console/build/CustomConsole.js:141:10)
      at QueueEvents.log (src/services/queue.service.ts:158:15)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-base.ts:121:20)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-events.ts:226:18)
      at QueueEvents.consumeEvents (node_modules/bullmq/src/classes/queue-events.ts:318:18)
      at QueueEvents.run (node_modules/bullmq/src/classes/queue-events.ts:272:9)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "⏳ Job 3852 is waiting to be processed.".

      156 |
      157 |     this.queueEvents.on('waiting', ({ jobId }) => {
    > 158 |       console.log(`⏳ Job ${jobId} is waiting to be processed.`);
          |               ^
      159 |     });
      160 |
      161 |     this.queueEvents.on('progress', ({ jobId, data }) => {

      at console.log (node_modules/@jest/console/build/CustomConsole.js:141:10)
      at QueueEvents.log (src/services/queue.service.ts:158:15)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-base.ts:121:20)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-events.ts:226:18)
      at QueueEvents.consumeEvents (node_modules/bullmq/src/classes/queue-events.ts:318:18)
      at QueueEvents.run (node_modules/bullmq/src/classes/queue-events.ts:272:9)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "⏳ Job 3853 is waiting to be processed.".

      156 |
      157 |     this.queueEvents.on('waiting', ({ jobId }) => {
    > 158 |       console.log(`⏳ Job ${jobId} is waiting to be processed.`);
          |               ^
      159 |     });
      160 |
      161 |     this.queueEvents.on('progress', ({ jobId, data }) => {

      at console.log (node_modules/@jest/console/build/CustomConsole.js:141:10)
      at QueueEvents.log (src/services/queue.service.ts:158:15)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-base.ts:121:20)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-events.ts:226:18)
      at QueueEvents.consumeEvents (node_modules/bullmq/src/classes/queue-events.ts:318:18)
      at QueueEvents.run (node_modules/bullmq/src/classes/queue-events.ts:272:9)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "⏳ Job 3854 is waiting to be processed.".

      156 |
      157 |     this.queueEvents.on('waiting', ({ jobId }) => {
    > 158 |       console.log(`⏳ Job ${jobId} is waiting to be processed.`);
          |               ^
      159 |     });
      160 |
      161 |     this.queueEvents.on('progress', ({ jobId, data }) => {

      at console.log (node_modules/@jest/console/build/CustomConsole.js:141:10)
      at QueueEvents.log (src/services/queue.service.ts:158:15)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-base.ts:121:20)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-events.ts:226:18)
      at QueueEvents.consumeEvents (node_modules/bullmq/src/classes/queue-events.ts:318:18)
      at QueueEvents.run (node_modules/bullmq/src/classes/queue-events.ts:272:9)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "⏳ Job 3855 is waiting to be processed.".

      156 |
      157 |     this.queueEvents.on('waiting', ({ jobId }) => {
    > 158 |       console.log(`⏳ Job ${jobId} is waiting to be processed.`);
          |               ^
      159 |     });
      160 |
      161 |     this.queueEvents.on('progress', ({ jobId, data }) => {

      at console.log (node_modules/@jest/console/build/CustomConsole.js:141:10)
      at QueueEvents.log (src/services/queue.service.ts:158:15)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-base.ts:121:20)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-events.ts:226:18)
      at QueueEvents.consumeEvents (node_modules/bullmq/src/classes/queue-events.ts:318:18)
      at QueueEvents.run (node_modules/bullmq/src/classes/queue-events.ts:272:9)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "⏳ Job 3856 is waiting to be processed.".

      156 |
      157 |     this.queueEvents.on('waiting', ({ jobId }) => {
    > 158 |       console.log(`⏳ Job ${jobId} is waiting to be processed.`);
          |               ^
      159 |     });
      160 |
      161 |     this.queueEvents.on('progress', ({ jobId, data }) => {

      at console.log (node_modules/@jest/console/build/CustomConsole.js:141:10)
      at QueueEvents.log (src/services/queue.service.ts:158:15)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-base.ts:121:20)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-events.ts:226:18)
      at QueueEvents.consumeEvents (node_modules/bullmq/src/classes/queue-events.ts:318:18)
      at QueueEvents.run (node_modules/bullmq/src/classes/queue-events.ts:272:9)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "⏳ Job 3857 is waiting to be processed.".

      156 |
      157 |     this.queueEvents.on('waiting', ({ jobId }) => {
    > 158 |       console.log(`⏳ Job ${jobId} is waiting to be processed.`);
          |               ^
      159 |     });
      160 |
      161 |     this.queueEvents.on('progress', ({ jobId, data }) => {

      at console.log (node_modules/@jest/console/build/CustomConsole.js:141:10)
      at QueueEvents.log (src/services/queue.service.ts:158:15)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-base.ts:121:20)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-events.ts:226:18)
      at QueueEvents.consumeEvents (node_modules/bullmq/src/classes/queue-events.ts:318:18)
      at QueueEvents.run (node_modules/bullmq/src/classes/queue-events.ts:272:9)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "⏳ Job 3858 is waiting to be processed.".

      156 |
      157 |     this.queueEvents.on('waiting', ({ jobId }) => {
    > 158 |       console.log(`⏳ Job ${jobId} is waiting to be processed.`);
          |               ^
      159 |     });
      160 |
      161 |     this.queueEvents.on('progress', ({ jobId, data }) => {

      at console.log (node_modules/@jest/console/build/CustomConsole.js:141:10)
      at QueueEvents.log (src/services/queue.service.ts:158:15)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-base.ts:121:20)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-events.ts:226:18)
      at QueueEvents.consumeEvents (node_modules/bullmq/src/classes/queue-events.ts:318:18)
      at QueueEvents.run (node_modules/bullmq/src/classes/queue-events.ts:272:9)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "⏳ Job 3859 is waiting to be processed.".

      156 |
      157 |     this.queueEvents.on('waiting', ({ jobId }) => {
    > 158 |       console.log(`⏳ Job ${jobId} is waiting to be processed.`);
          |               ^
      159 |     });
      160 |
      161 |     this.queueEvents.on('progress', ({ jobId, data }) => {

      at console.log (node_modules/@jest/console/build/CustomConsole.js:141:10)
      at QueueEvents.log (src/services/queue.service.ts:158:15)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-base.ts:121:20)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-events.ts:226:18)
      at QueueEvents.consumeEvents (node_modules/bullmq/src/classes/queue-events.ts:318:18)
      at QueueEvents.run (node_modules/bullmq/src/classes/queue-events.ts:272:9)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "⏳ Job 3860 is waiting to be processed.".

      156 |
      157 |     this.queueEvents.on('waiting', ({ jobId }) => {
    > 158 |       console.log(`⏳ Job ${jobId} is waiting to be processed.`);
          |               ^
      159 |     });
      160 |
      161 |     this.queueEvents.on('progress', ({ jobId, data }) => {

      at console.log (node_modules/@jest/console/build/CustomConsole.js:141:10)
      at QueueEvents.log (src/services/queue.service.ts:158:15)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-base.ts:121:20)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-events.ts:226:18)
      at QueueEvents.consumeEvents (node_modules/bullmq/src/classes/queue-events.ts:318:18)
      at QueueEvents.run (node_modules/bullmq/src/classes/queue-events.ts:272:9)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "⏳ Job 3861 is waiting to be processed.".

      156 |
      157 |     this.queueEvents.on('waiting', ({ jobId }) => {
    > 158 |       console.log(`⏳ Job ${jobId} is waiting to be processed.`);
          |               ^
      159 |     });
      160 |
      161 |     this.queueEvents.on('progress', ({ jobId, data }) => {

      at console.log (node_modules/@jest/console/build/CustomConsole.js:141:10)
      at QueueEvents.log (src/services/queue.service.ts:158:15)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-base.ts:121:20)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-events.ts:226:18)
      at QueueEvents.consumeEvents (node_modules/bullmq/src/classes/queue-events.ts:318:18)
      at QueueEvents.run (node_modules/bullmq/src/classes/queue-events.ts:272:9)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "⏳ Job 3862 is waiting to be processed.".

      156 |
      157 |     this.queueEvents.on('waiting', ({ jobId }) => {
    > 158 |       console.log(`⏳ Job ${jobId} is waiting to be processed.`);
          |               ^
      159 |     });
      160 |
      161 |     this.queueEvents.on('progress', ({ jobId, data }) => {

      at console.log (node_modules/@jest/console/build/CustomConsole.js:141:10)
      at QueueEvents.log (src/services/queue.service.ts:158:15)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-base.ts:121:20)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-events.ts:226:18)
      at QueueEvents.consumeEvents (node_modules/bullmq/src/classes/queue-events.ts:318:18)
      at QueueEvents.run (node_modules/bullmq/src/classes/queue-events.ts:272:9)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "⏳ Job 3863 is waiting to be processed.".

      156 |
      157 |     this.queueEvents.on('waiting', ({ jobId }) => {
    > 158 |       console.log(`⏳ Job ${jobId} is waiting to be processed.`);
          |               ^
      159 |     });
      160 |
      161 |     this.queueEvents.on('progress', ({ jobId, data }) => {

      at console.log (node_modules/@jest/console/build/CustomConsole.js:141:10)
      at QueueEvents.log (src/services/queue.service.ts:158:15)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-base.ts:121:20)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-events.ts:226:18)
      at QueueEvents.consumeEvents (node_modules/bullmq/src/classes/queue-events.ts:318:18)
      at QueueEvents.run (node_modules/bullmq/src/classes/queue-events.ts:272:9)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "⏳ Job 3864 is waiting to be processed.".

      156 |
      157 |     this.queueEvents.on('waiting', ({ jobId }) => {
    > 158 |       console.log(`⏳ Job ${jobId} is waiting to be processed.`);
          |               ^
      159 |     });
      160 |
      161 |     this.queueEvents.on('progress', ({ jobId, data }) => {

      at console.log (node_modules/@jest/console/build/CustomConsole.js:141:10)
      at QueueEvents.log (src/services/queue.service.ts:158:15)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-base.ts:121:20)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-events.ts:226:18)
      at QueueEvents.consumeEvents (node_modules/bullmq/src/classes/queue-events.ts:318:18)
      at QueueEvents.run (node_modules/bullmq/src/classes/queue-events.ts:272:9)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "⏳ Job 3865 is waiting to be processed.".

      156 |
      157 |     this.queueEvents.on('waiting', ({ jobId }) => {
    > 158 |       console.log(`⏳ Job ${jobId} is waiting to be processed.`);
          |               ^
      159 |     });
      160 |
      161 |     this.queueEvents.on('progress', ({ jobId, data }) => {

      at console.log (node_modules/@jest/console/build/CustomConsole.js:141:10)
      at QueueEvents.log (src/services/queue.service.ts:158:15)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-base.ts:121:20)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-events.ts:226:18)
      at QueueEvents.consumeEvents (node_modules/bullmq/src/classes/queue-events.ts:318:18)
      at QueueEvents.run (node_modules/bullmq/src/classes/queue-events.ts:272:9)


  ●  Cannot log after tests are done. Did you forget to wait for something async in your test?
    Attempted to log "⏳ Job 3866 is waiting to be processed.".

      156 |
      157 |     this.queueEvents.on('waiting', ({ jobId }) => {
    > 158 |       console.log(`⏳ Job ${jobId} is waiting to be processed.`);
          |               ^
      159 |     });
      160 |
      161 |     this.queueEvents.on('progress', ({ jobId, data }) => {

      at console.log (node_modules/@jest/console/build/CustomConsole.js:141:10)
      at QueueEvents.log (src/services/queue.service.ts:158:15)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-base.ts:121:20)
      at QueueEvents.emit (node_modules/bullmq/src/classes/queue-events.ts:226:18)
      at QueueEvents.consumeEvents (node_modules/bullmq/src/classes/queue-events.ts:318:18)
      at QueueEvents.run (node_modules/bullmq/src/classes/queue-events.ts:272:9)

ShaySarussiElshten reviewed (changes requested):
@dori654 
you saw my meesage ?
the test didn׳t run

Shaybush commented on docker-compose.yml:
```diff
@@ -36,6 +38,12 @@ services:
     volumes:
       - mongo_data:/data/db
 
+  redis:  
+    image: redis:latest
+    container_name: redis
+    ports:
+      - "6379:6379"
+
```
add volume

Shaybush commented on src/services/queue.service.ts:
```diff
@@ -0,0 +1,177 @@
+  private queue: Queue;
+  private worker: Worker;
+  private queueEvents: QueueEvents;
+
+  constructor() {
+    const redisOptions: QueueOptions = {
```
move to private above

dori654 replied:
cant do, becase the issue is that InversifyJS expects a class with a public constructor to allow dependency injection.
I can remove injectable and it will work

Shaybush commented on src/services/queue.service.ts:
```diff
@@ -0,0 +1,177 @@
+    
+  }
+
+  public getQueueEvents() {
+    return this.queueEvents; // ✅ Exposes queue events
+  }
```
delete redundant comments
```suggestion
  public getQueueEvents() {
    return this.queueEvents;
  }
```

Shaybush commented on src/services/queue.service.ts:
```diff
@@ -0,0 +1,177 @@
+   * @param data - The job data.
+   * @param delay - Optional delay in milliseconds.
+   * @param priority - Job priority (lower number = higher priority).
+   */
+  public async enqueue(data: any, delay = 0, priority = 1): Promise<Job> {
+    return await this.queue.add('job', data, {
+      attempts: config.maxRetries,
+      backoff: { type: 'fixed', delay: config.retryDelay }, 
+      delay,
+      priority
+    });
+  }
```
```suggestion
   */
  public async enqueue<T>(data: T, delay = 0, priority = 1): Promise<Job> {
    return await this.queue.add('job', data, {
      attempts: config.maxRetries,
      backoff: { type: 'fixed', delay: config.retryDelay }, 
      delay,
      priority
    });
  }
```

Shaybush commented on src/services/queue.service.ts:
```diff
@@ -0,0 +1,177 @@
+   * @param jobId - The ID of the job.
+   * @returns JobState or null if not found.
+   */
+  public async getJobStatus(jobId: string): Promise<JobState | null> {
+    const job = await this.queue.getJob(jobId);
+    if (!job) return null;
```
2 things 
1 - please log this error 
2 - throw error to be handled in the catch (more information to the user)

Shaybush commented on src/services/queue.service.ts:
```diff
@@ -0,0 +1,177 @@
+    try {
+      await job.remove();
+      return true;
+    } catch (error) {
+      console.error(`❌ Could not remove job ${jobId}:`, error);
+      return false;
+    }
```
log it, and throw error

Shaybush commented on src/services/queue.service.ts:
```diff
@@ -0,0 +1,177 @@
+  private createWorker(): Worker {
+    return new Worker(
+      config.queueName,
+      async (job) => {
+        console.log(`🔄 Processing job ${job.id} with data:`, job.data);
```
use logger

Shaybush commented on src/services/queue.service.ts:
```diff
@@ -0,0 +1,177 @@
+      config.queueName,
+      async (job) => {
+        console.log(`🔄 Processing job ${job.id} with data:`, job.data);
+        await new Promise(resolve => setTimeout(resolve, config.pollInterval));
```
please check it

dori654 committed:
Merge https://github.com/InnovatorsHub1/backend into BE-33

ShaySarussiElshten committed:
in

ShaySarussiElshten committed:
fix problam docker + doker compose and merge

ShaySarussiElshten commented:
Hey @dori654 ,
i fixed some issues related to Docker and dockerfile , and the merge you did to master wasn’t done properly.
Please note that you still have tests that are failing.

dori654 committed:
Refactor QueueService to enhance job handling and retry logic

- Changed queue property from private to protected for better accessibility.
- Improved enqueue method to handle job options more flexibly, including support for jobs that should fail.
- Updated dequeue method to recognize new job states.
- Modified retryJob method to create a new job upon retrying a failed job, preserving the original job's data.
- Added a new retry method for direct job retries.
- Enhanced workers.queue.ts to process fail tasks correctly.
- Updated unit tests for better structure and added scenarios for job retries and concurrent processing.
